File: lab1/FinderOfCopy.java
import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.MulticastSocket;
import java.nio.charset.StandardCharsets;
import java.util.HashSet;
import java.util.Set;

public class FinderOfCopy {
    public final int PORT = 10000;
    private DatagramSocket datagramSocket = null;
    private MulticastSocket multicastSocket = null;
    private byte[] buf = new byte[256];
    private String myIP;
    private InetAddress inetAddress;
    private Set <Pair<String, String>> machineSet= new HashSet<>();


    public FinderOfCopy(String ip) throws Exception {
        find(ip);
    }

    public void find(String ip) throws Exception {
        myIP = InetAddress.getLocalHost().getHostAddress();

        inetAddress = InetAddress.getByName(ip);
        datagramSocket = new DatagramSocket();
        multicastSocket = new MulticastSocket(PORT);
        multicastSocket.joinGroup(inetAddress);

        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            String sendBuf = "DISCONNECT " + myIP + " " + ProcessHandle.current().pid();
            DatagramPacket packet = new DatagramPacket(sendBuf.getBytes(StandardCharsets.UTF_8), sendBuf.length(), inetAddress, PORT);
            try {
                datagramSocket.send(packet);
            } catch (IOException ignored) {
            }
        }));

        start(inetAddress);
    }

    private void start(InetAddress inetAddress) throws IOException {
        connect();
        datagramSocket.setBroadcast(true);
        while (true) {
            receiveMsg();
        }
    }

    private void connect() throws IOException {
        String sendBuf = "CONNECT " + myIP + " " + ProcessHandle.current().pid();
        DatagramPacket packet = new DatagramPacket(sendBuf.getBytes(StandardCharsets.UTF_8), sendBuf.length(), inetAddress, PORT);
        datagramSocket.send(packet);
    }


    private void receiveMsg() throws IOException {
        DatagramPacket packet = new DatagramPacket(buf, buf.length);
        multicastSocket.receive(packet);
        msgHandler(packet);
    }


    public void msgHandler(DatagramPacket packet) throws IOException {
        String[] dataArgs = new String(packet.getData(), 0, packet.getLength()).split(" +");
        switch (dataArgs[0]) {
            case "CONNECT": {
                if (!dataArgs[2].equals(String.valueOf(ProcessHandle.current().pid())) || !dataArgs[1].equals(String.valueOf(myIP))) {
                    System.out.println(" + " + dataArgs[1] + " " + dataArgs[2]);
                    machineSet.add(new Pair<>(dataArgs[1], dataArgs[2]));
                    answer();
                }

                break;
            }
            case "OLD_MACHINE": {
                if (!dataArgs[2].equals(String.valueOf(ProcessHandle.current().pid())) || !dataArgs[1].equals(String.valueOf(myIP))) {
                    if(!machineSet.contains(new Pair<>(dataArgs[1], dataArgs[2]))){
                        machineSet.add(new Pair<>(dataArgs[1], dataArgs[2]));
                        System.out.println(" * " + dataArgs[1] + " " + dataArgs[2]);
                    }
                }

                break;
            }
            case "DISCONNECT": {
                if (!dataArgs[2].equals(String.valueOf(ProcessHandle.current().pid())) || !dataArgs[1].equals(String.valueOf(myIP))) {

                    if(machineSet.contains(new Pair<>(dataArgs[1], dataArgs[2]))){
                        machineSet.remove(new Pair<>(dataArgs[1], dataArgs[2]));
                        System.out.println(" - " + dataArgs[1] + " " + dataArgs[2]);
                    }
                }

                break;
            }
        }
    }

    public void answer() throws IOException {
        String sendBuf = "OLD_MACHINE " + myIP + " " + ProcessHandle.current().pid();
        DatagramPacket packet = new DatagramPacket(sendBuf.getBytes(StandardCharsets.UTF_8), sendBuf.length(), inetAddress, PORT);
        datagramSocket.send(packet);
    }

    private enum InstanceMsg {
        CONNECT,
        OLD_MACHINE,
        DISCONNECT;

        @Override
        public String toString() {
            return switch (this) {
                case CONNECT -> "CONNECT";
                case OLD_MACHINE -> "OLD_MACHINE";
                case DISCONNECT -> "DISCONNECT";
            };
        }
    }
}

File: lab1/src/main/kotlin/FinderOfCopy.kt
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.net.MulticastSocket
import java.nio.charset.StandardCharsets
import java.util.*
import kotlin.math.abs

class FinderOfCopy(
    private val date: Date = Date(),
    private val hashMap: HashMap<Node, Long> = HashMap<Node, Long>(),
    private val myIp: String? = InetAddress.getLocalHost().hostAddress,
    private val PORT: Int = 10000,
    private val pid: Long = ProcessHandle.current().pid(),
) {

    fun find(ip: String) {
        val datagramSocket = DatagramSocket()
        val multicastSocket = MulticastSocket(PORT)
        val inetAddress = InetAddress.getByName(ip)
        var byteArray = ByteArray(256)
        multicastSocket.soTimeout = 10000
        multicastSocket.joinGroup(inetAddress)
        datagramSocket.broadcast = true
        var timeLastSend: Long = 0
        println("My info: $myIp $pid");
        while (true) {
            //recv data
            var datagramPacketRecv = DatagramPacket(byteArray, byteArray.size)
            try {
                multicastSocket.receive(datagramPacketRecv)
                msgHandler(datagramPacketRecv)
            } catch (ex: Exception) {
            }
            //send data
            if (abs(timeLastSend - date.time) > 10000 || timeLastSend.equals(0)) {
                val buffer = "$myIp $pid"

                val datagramPacketSend =
                    DatagramPacket(buffer.toByteArray(StandardCharsets.UTF_8), buffer.length, inetAddress, PORT)
                datagramSocket.send(datagramPacketSend)
            } else {
                if (timeLastSend.equals(0)) timeLastSend = date.time
            }
            if (abs(timeLastSend - date.time) > 60000) {
                healthChecker()
            }
        }
    }

    private fun msgHandler(packet: DatagramPacket) {
        var args = String(packet.data, 0, packet.length, StandardCharsets.UTF_8).split(" ").toTypedArray()
        if (!args[1].equals(pid.toString()) || !args[0].equals(myIp)) { //self checker
            if (hashMap[Node(args[0], args[1])] == null) {
                hashMap[Node(args[0], args[1])] = date.time //add new copy
                val ipAdd = args[0]
                val pidAdd = args[1]
                println("+ $ipAdd $pidAdd")
            } else {
                hashMap[Node(args[0], args[1])] = date.time //update time
            }
        }
    }

    private fun healthChecker() {
        for (it in hashMap.toList().iterator()) {
            if (abs(it.second - date.time) > 100000) {
                hashMap.remove(it.first)
                val ipRemove = it.first
                val pidRemove = it.second
                println("- $ipRemove $pidRemove")
            }
        }
    }
}